tests/python/pants_test/engine/test_isolated_process.py

```diff
@@ -84,22 +154,125 @@ def java_sources_to_javac_args(sources_snapshot, out_dir):
                                    directories_to_create=(out_dir.path,))
 
 
-def process_request_from_java_sources(binary):
-  env = []
-  return ExecuteProcessRequest(
-    argv=binary.prefix_of_command() + ('-version',),
-    env=env,
-    input_files_digest="e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
+class JavacVersionCommand(Javac):
+
+  def gen_argv(self, snapshots=None):
+    if snapshots:
+      raise ValueError("JavacVersionCommand cannot use input snapshots '{}'"
+                       .format(snapshots))
+    return (self.bin_path, '-version',)
+
+
+EMPTY_FINGERPRINT = 'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855'
+
+
+@rule(ExecuteProcessRequest, [Select(JavacVersionCommand)])
+def process_request_from_javac_version(javac_version_command):
+  yield ExecuteProcessRequest(
+    argv=javac_version_command.gen_argv(),
+    env=[],
+    input_files_digest=EMPTY_FINGERPRINT,
     digest_length=0)
 
 
-def process_request_java_args_from_java_sources(binary, sources_snapshot, out_dir):
-  env = []
-  return ExecuteProcessRequest(
-    argv=binary.prefix_of_command() + tuple(f.path for f in sources_snapshot.files),
-    input_files_digest=str(sources_snapshot.fingerprint),
-    digest_length=sources_snapshot.digest_length,
-    env=env)
+class JavacVersionOutput(datatype('JavacVersionOutput', [
+    'exit_code',
+    'version_output',
+])): pass
+
+
+@rule(JavacVersionOutput, [Select(JavacVersionCommand)])
+def get_javac_version_output(javac_version_command):
+  javac_version_proc_req = yield Get(
+    ExecuteProcessRequest, JavacVersionCommand, javac_version_command)
+  javac_version_proc_result = yield Get(
+    ExecuteProcessResult, ExecuteProcessRequest, javac_version_proc_req)
+  yield JavacVersionOutput(
+    exit_code=javac_version_proc_result.exit_code,
+    version_output=javac_version_proc_result.stderr,
+  )
+
+
+class JavacCompileCommand(Javac):
+
+  def gen_argv(self, snapshots):
+    snapshot_file_paths = []
```

Comment:
Probably wants to deduplicate; use an `OrderedSet`?

Reply:
Made the change and left a TODO here because deduplication of file paths sounds like something we should handle for users of `@rule`s and `Binary`s.